File: typescript/src/bittorrent/logic.ts
// SPDX-License-Identifier: MIT
// Rarest-first picker + SHA-1-verified piece store.
import { createHash } from 'node:crypto';
import { Bitfield } from './wire.js';

export class RarestFirstPicker {
  private have: boolean[];
  private inflight: boolean[];
  private avail: number[];
  private peerHas = new Map<string, boolean[]>();

  constructor(private count: number) {
    this.have = new Array(count).fill(false);
    this.inflight = new Array(count).fill(false);
    this.avail = new Array(count).fill(0);
  }

  setHave(i: number): void {
    if (i >= 0 && i < this.count) {
      this.have[i] = true;
      this.inflight[i] = false;
    }
  }

  addPeer(peer: string): void {
    if (!this.peerHas.has(peer)) this.peerHas.set(peer, new Array(this.count).fill(false));
  }

  peerHasPiece(peer: string, i: number): void {
    this.addPeer(peer);
    const has = this.peerHas.get(peer) as boolean[];
    if (i >= 0 && i < this.count && !has[i]) {
      has[i] = true;
      this.avail[i]++;
    }
  }

  pickFor(peer: string): number {
    const has = this.peerHas.get(peer);
    if (!has) return -1;
    let best = -1;
    let bestAvail = 0;
    for (let i = 0; i < this.count; i++) {
      if (this.have[i] || this.inflight[i] || !has[i]) continue;
      if (best === -1 || this.avail[i] < bestAvail) {
        best = i;
        bestAvail = this.avail[i];
      }
    }
    if (best !== -1) this.inflight[best] = true;
    return best;
  }

  release(i: number): void {
    if (i >= 0 && i < this.count) this.inflight[i] = false;
  }

  isComplete(): boolean {
    return this.count > 0 && this.have.every((h) => h);
  }
}

export class PieceStore {
  private pieces = new Map<number, Buffer>();

  constructor(
    private pieceLength: number,
    private totalLength: number,
    public pieceHashes: Buffer[],
  ) {}

  pieceCount(): number {
    return this.pieceHashes.length;
  }

  lengthOfPiece(i: number): number {
    if (i < 0 || i >= this.pieceHashes.length) return 0;
    if (i === this.pieceHashes.length - 1) return this.totalLength - i * this.pieceLength;
    return this.pieceLength;
  }

  has(i: number): boolean {
    return this.pieces.has(i);
  }

  tryComplete(i: number, data: Buffer): boolean {
    if (i < 0 || i >= this.pieceHashes.length) return false;
    if (data.length !== this.lengthOfPiece(i)) return false;
    if (!createHash('sha1').update(data).digest().equals(this.pieceHashes[i])) return false;
    this.pieces.set(i, Buffer.from(data));
    return true;
  }

  readBlock(i: number, begin: number, length: number): Buffer | null {
    const p = this.pieces.get(i);
    if (!p || begin < 0 || begin + length > p.length) return null;
    return Buffer.from(p.subarray(begin, begin + length));
  }

  buildBitfield(): Bitfield {
    const bf = new Bitfield(this.pieceHashes.length);
    for (let i = 0; i < this.pieceHashes.length; i++) if (this.has(i)) bf.set(i);
    return bf;
  }

  isComplete(): boolean {
    return this.pieces.size === this.pieceHashes.length;
  }

  assemble(): Buffer | null {
    if (!this.isComplete()) return null;
    const parts: Buffer[] = [];
    for (let i = 0; i < this.pieceHashes.length; i++) parts.push(this.pieces.get(i) as Buffer);
    return Buffer.concat(parts);
  }
}

export function pieceStoreFromContent(data: Buffer, pieceLength: number): PieceStore {
  const pieceCount = Math.ceil(data.length / pieceLength);
  const hashes: Buffer[] = [];
  const store = new PieceStore(pieceLength, data.length, []);
  for (let i = 0; i < pieceCount; i++) {
    const start = i * pieceLength;
    const end = Math.min(start + pieceLength, data.length);
    hashes.push(createHash('sha1').update(data.subarray(start, end)).digest());
    (store as unknown as { pieces: Map<number, Buffer> }).pieces.set(i, Buffer.from(data.subarray(start, end)));
  }
  store.pieceHashes = hashes;
  return store;
}

File: rust/src/bittorrent/logic.rs
// SPDX-License-Identifier: MIT
//! Rarest-first picker + SHA-1-verified piece store.

use sha1::{Digest, Sha1};
use std::collections::HashMap;

pub struct RarestFirstPicker {
    count: usize,
    have: Vec<bool>,
    inflight: Vec<bool>,
    avail: Vec<usize>,
    peer_has: HashMap<String, Vec<bool>>,
}

impl RarestFirstPicker {
    pub fn new(count: usize) -> Self {
        Self { count, have: vec![false; count], inflight: vec![false; count], avail: vec![0; count], peer_has: HashMap::new() }
    }
    pub fn set_have(&mut self, i: usize) {
        if i < self.count {
            self.have[i] = true;
            self.inflight[i] = false;
        }
    }
    pub fn add_peer(&mut self, peer: &str) {
        self.peer_has.entry(peer.to_string()).or_insert_with(|| vec![false; self.count]);
    }
    pub fn peer_has_piece(&mut self, peer: &str, i: usize) {
        self.add_peer(peer);
        if i < self.count {
            let h = self.peer_has.get_mut(peer).unwrap();
            if !h[i] {
                h[i] = true;
                self.avail[i] += 1;
            }
        }
    }
    pub fn pick_for(&mut self, peer: &str) -> i64 {
        let has = match self.peer_has.get(peer) {
            Some(h) => h.clone(),
            None => return -1,
        };
        let mut best: i64 = -1;
        let mut best_avail = 0usize;
        for i in 0..self.count {
            if self.have[i] || self.inflight[i] || !has[i] {
                continue;
            }
            if best == -1 || self.avail[i] < best_avail {
                best = i as i64;
                best_avail = self.avail[i];
            }
        }
        if best >= 0 {
            self.inflight[best as usize] = true;
        }
        best
    }
    pub fn release(&mut self, i: usize) {
        if i < self.count {
            self.inflight[i] = false;
        }
    }
    pub fn is_complete(&self) -> bool {
        self.count > 0 && self.have.iter().all(|&h| h)
    }
}

pub struct PieceStore {
    piece_length: usize,
    total_length: usize,
    pub piece_hashes: Vec<Vec<u8>>,
    pieces: HashMap<usize, Vec<u8>>,
}

impl PieceStore {
    pub fn new(piece_length: usize, total_length: usize, piece_hashes: Vec<Vec<u8>>) -> Self {
        Self { piece_length, total_length, piece_hashes, pieces: HashMap::new() }
    }
    pub fn piece_count(&self) -> usize {
        self.piece_hashes.len()
    }
    pub fn length_of_piece(&self, i: usize) -> usize {
        if i >= self.piece_hashes.len() {
            return 0;
        }
        if i == self.piece_hashes.len() - 1 {
            self.total_length - i * self.piece_length
        } else {
            self.piece_length
        }
    }
    pub fn has(&self, i: usize) -> bool {
        self.pieces.contains_key(&i)
    }
    pub fn try_complete(&mut self, i: usize, data: &[u8]) -> bool {
        if i >= self.piece_hashes.len() || data.len() != self.length_of_piece(i) {
            return false;
        }
        let mut h = Sha1::new();
        h.update(data);
        if h.finalize().as_slice() != self.piece_hashes[i].as_slice() {
            return false;
        }
        self.pieces.insert(i, data.to_vec());
        true
    }
    pub fn is_complete(&self) -> bool {
        self.pieces.len() == self.piece_hashes.len()
    }
    pub fn assemble(&self) -> Option<Vec<u8>> {
        if !self.is_complete() {
            return None;
        }
        let mut out = Vec::with_capacity(self.total_length);
        for i in 0..self.piece_hashes.len() {
            out.extend_from_slice(&self.pieces[&i]);
        }
        Some(out)
    }
}

pub fn piece_store_from_content(data: &[u8], piece_length: usize) -> PieceStore {
    let pc = data.len().div_ceil(piece_length);
    let mut hashes = Vec::new();
    let mut store = PieceStore::new(piece_length, data.len(), Vec::new());
    for i in 0..pc {
        let start = i * piece_length;
        let end = (start + piece_length).min(data.len());
        let mut h = Sha1::new();
        h.update(&data[start..end]);
        hashes.push(h.finalize().to_vec());
        store.pieces.insert(i, data[start..end].to_vec());
    }
    store.piece_hashes = hashes;
    store
}
